### ISSUE TITLE
Proposal: New argument passing convention

### ISSUE THREAD

**Issue by peterbell10** [Quansight-Labs/uarray#164]
## Adding new arguments to functions
**Problem**:
Currently, `__ua_function__` takes its arguments by the standard python `args, kwargs` pair. However, I think this creates a problem under maintenance. To illustrate this, consider that our backend supports a simple `fft` function
```python
def fft(x)
```
and in a new version of scipy, we want to add support for different normalization modes via a `norm` argument
```python
def fft(x, norm=None)
```

Because the argument has been added in a backwards compatible way, it's still fine to call `fft(x)` on the old backend which doesn't support the `norm` argument. Because of #162, it's also fine to explicitly call `fft(x, norm=None)`. However `fft(x, None)` doesn't work because positional arguments are not canonicalized.

**Solution**: convert all arguments into one dict and canonicalize all of them.

## Adding hint arguments
As an extension, we can also consider another kind of argument
```python
def fft(x, norm=None, overwrite_x=False)
```
in `scipy.fft`, `overwrite_x` is a hint to the implementation that it can reuse the memory from the input array `x`. Importantly, there's no obligation for the implementation to actually respect the argument and it doesn't effect the result of the function.

In other words, its actually okay to use a backend that doesn't know about this argument, even if called with `fft(x, overwrite_x=True)`.

So, I propose spitting the argument dictionary into two dicts: `args` and `hints`. If the backend sees a key in `args` that it doesn't recognise, it should return `NotImplemented`, but any unknown `hints` can be safely ignored.

**Comment by hameerabbasi** [comment]
> Solution: convert all arguments into one dict and canonicalize all of them.

While this is the common case, there are positional-only and keyword-only arguments... So keeping `args` around is needed. One dict won't work, but we can move every _possible_ arg into the `dict`.

**Comment by peterbell10** [comment]
Positional only arguments still have names so I don't see why they couldn't be inserted into a `dict`.

**Comment by hameerabbasi** [comment]
They do... But they couldn't be called in the form `f(**kwargs)`.

**Comment by peterbell10** [comment]
Positional-only arguments only matter for the _caller-facing_ signature (`argument_extractor`) so there's no reason why the _backend_ couldn't be callable as `f(**kwargs)`.

**Comment by hameerabbasi** [comment]
> Positional-only arguments only matter for the _caller-facing_ signature (`argument_extractor`) so there's no reason why the _backend_ couldn't be callable as `f(**kwargs)`.

Ah, excellent point. We'd have to document this, though, make sure all backends don't have positional-only arguments and that API multimethods have sensible names.

**Comment by hameerabbasi** [comment]
In that light, I'm in favor of the first but not in favour of the second... How would you "mark" something as a hint argument?

**Comment by peterbell10** [comment]
>How would you "mark" something as a hint argument?

As an example, I could pass a list of their names to `generate_multimethod`

>I would instead prefer to drop defaults than accept incorrect behavior.

But hints shouldn't effect the result, so there is no "incorrect behavior".

**Comment by hameerabbasi** [comment]
Ah, so things that are purely performance-related.

**Comment by hameerabbasi** [comment]
Are you willing to make a PR?

**Comment by peterbell10** [comment]
Sure. Should I base it off of #160?

**Comment by hameerabbasi** [comment]
I'll merge #160 now after CI passes, then we can start on this. Thanks for being so willing to contribute!

**Comment by hameerabbasi** [comment]
@peterbell10 Do you think hint arguments would still help and not bring up overhead too much?

**Comment by peterbell10** [comment]
Depending on the  design to extract those arguemtns it might not be too high of an overhead. However, I also think it would be tricky to mix hint args with positional args. I'd be okay with just closing this as the use cases are fairly rare anyway.

**Comment by hameerabbasi** [comment]
We'll reopen this if there's a need for it.
